lib/writers/git.go

```diff
@@ -68,125 +68,139 @@ func NewGitWriter(logger logr.Logger, stateStoreSpec platformv1alpha1.GitStateSt
+		return "", nil, nil, err
 	}
+	return localTmpDir, repo, worktree, nil
+}
 
-	workTreeFilePath := filepath.Join(g.path, fileName)
-	absoluteFilePath := filepath.Join(localTmpDir, workTreeFilePath)
-
-	//We need to protect against paths containg `..`
-	//filepath.Join expands any '../' in the path to the actual, e.g. /tmp/foo/../ resolves to /tmp/
-	//To ensure they can't write to files on disk outside of the tmp git repostiroy we check the absolute path
-	//returned by `filepath.Join` is still contained with the git repository:
-	// Note: This means `../` can still be used, but only if the end result is still contained within the git repository
-	if !strings.HasPrefix(absoluteFilePath, localTmpDir) {
-		log.Error(nil, "path of file to write is not located within the git repostiory", "absolutePath", absoluteFilePath, "tmpDir", localTmpDir, "repoPath", workTreeFilePath, "path", g.path)
-		return nil //We don't want to retry as this isn't a recoverable error. Log error and return nil.
-	}
+func (g *GitWriter) WriteDirWithObjects(deleteExistingContentsInDir bool, subDir string, toWrite ...platformv1alpha1.Workload) error {
+	dirInGitRepo := filepath.Join(g.path, subDir)
+	logger := g.Log.WithValues(
+		"dir", dirInGitRepo,
+		"branch", g.gitServer.Branch,
+	)
 
-	if os.MkdirAll(filepath.Dir(absoluteFilePath), 0700); err != nil {
-		log.Error(err, "could not generate local directories")
-		return err
+	if len(toWrite) == 0 {
+		logger.Info("Empty workloads. Nothing to write to Git")
+		return nil
 	}
 
-	if err := ioutil.WriteFile(absoluteFilePath, toWrite, 0644); err != nil {
-		log.Error(err, "could not write to file")
+	localTmpDir, repo, worktree, err := g.setupLocalDirectoryWithRepo(logger)
+	if err != nil {
 		return err
 	}
+	defer os.RemoveAll(filepath.Dir(localTmpDir))
 
-	if _, err := worktree.Add(workTreeFilePath); err != nil {
-		log.Error(err, "could not add file to worktree")
-		return err
+	if deleteExistingContentsInDir {
+		logger.Info("checking if any existing directories needs to be deleted")
+		if _, err := worktree.Filesystem.Lstat(dirInGitRepo); err == nil {
+			logger.Info("deleting existing content")
+			if _, err := worktree.Remove(dirInGitRepo); err != nil {
+				logger.Error(err, "could not add directory deletion to worktree", "dir", dirInGitRepo)
+				return err
+			}
+		}
 	}
 
-	if err := g.commitAndPush(repo, worktree, Add, workTreeFilePath, log); err != nil {
-		return err
-	}
+	var filesCommitted []string
+	for _, item := range toWrite {
+		//worker-cluster/resources/<rr-namespace>/<promise-name>/<rr-name>/foo/bar/baz.yaml
+		worktreeFilePath := filepath.Join(dirInGitRepo, item.Filepath)
+		logger := logger.WithValues(
+			"filepath", worktreeFilePath,
+		)
+
+		///tmp/git-dir/worker-cluster/resources/<rr-namespace>/<promise-name>/<rr-name>/foo/bar/baz.yaml
+		absoluteFilePath := filepath.Join(localTmpDir, worktreeFilePath)
+
+		//We need to protect against paths containg `..`
```

Comment:
are we protecting against ourselves here? 🤔

Reply:
I think anyone who creates works/workplacements can pass in `..` so its a reasonable check (this was pre-existing before this PR)